Switch from gometalinter to golangci-lint

mjtrangoni opened the pull request:
Hi @torkelo @xlson @bergquist ,

I found [golangci-lint](https://github.com/golangci/golangci-lint), which is on average [5 times faster](https://github.com/golangci/golangci-lint#performance) than [gometalinter](https://github.com/alecthomas/gometalinter).

This fixes some issues found by **golangci-lint**, mostly **goconst** issues which are trivial.
I leave **staticcheck** running with gometalinter as there is an issue with it, see [#369](https://github.com/golangci/golangci-lint/issues/369).

IMHO, this add value to the linter checks, as it is running quickly and smartly.

See, i.e. running **all** linters except staticcheck takes 11s!!

```
$ time golangci-lint run --print-issued-lines=false --enable-all --deadline 10m --disable=staticcheck | awk '{print $NF}'| sort | uniq -c
     50 (dupl)
     50 (errcheck)
     50 (gochecknoglobals)
      3 (gochecknoinits)
     40 (gocritic)
      7 (gocyclo)
      3 (goimports)
     50 (golint)
     20 (gosec)
      4 (interfacer)
     50 (lll)
     28 (maligned)
      5 (misspell)
     11 (prealloc)
     15 (scopelint)
      3 (stylecheck)
      2 (typecheck)
     50 (unparam)

real	0m11.028s
user	0m31.194s
sys	0m2.745s
```

And here what we would be running with this PR,

```
$ time golangci-lint run --deadline 10m --disable-all --enable=goconst --enable=deadcode --enable=gofmt --enable=ineffassign --enable=structcheck --enable=unconvert --enable=varcheck

real	0m3.134s
user	0m4.476s
sys	0m0.889s
```

mjtrangoni committed:
pkg/setting/setting.go: Fix goconst issues.

See,
$ golangci-lint run --disable-all --deadline 10m --enable=goconst
pkg/setting/setting.go:249:30: string `windows` has 5 occurrences, make it a constant (goconst)
	IsWindows = runtime.GOOS == "windows"
	                            ^

mjtrangoni committed:
pkg/services/alerting/notifiers/*: Fix goconst issue with empty JSON string.

See,
$ golangci-lint run --disable-all --deadline 10m --enable=goconst
pkg/services/alerting/notifiers/alertmanager_test.go:63:13: string `{ }` has 18 occurrences, make it a constant (goconst)
				json := `{ }`
				        ^

mjtrangoni committed:
pkg/services/alerting/notifiers/*: Fix goconst issue with google URL in JSON string.

See,
$ golangci-lint run --disable-all --deadline 10m --enable=goconst
pkg/services/alerting/notifiers/hipchat_test.go:30:13: string `
				{
          url: "http://google.com"
				}` has 6 occurrences, make it a constant (goconst)
				json := `
				        ^

mjtrangoni committed:
middleware_test.go: fix goconst issue with X-WEBAUTH-USER.

See,
$ golangci-lint run --disable-all --deadline 10m --enable=goconst
pkg/middleware/middleware_test.go:269:34: string `X-WEBAUTH-USER` has 11 occurrences, make it a constant (goconst)
			setting.AuthProxyHeaderName = "X-WEBAUTH-USER"
			                              ^

mjtrangoni committed:
pkg/middleware/*: Fix goconst issue with `username` string

See,
$ golangci-lint run --disable-all --deadline 10m --enable=goconst
pkg/middleware/auth_proxy.go:107:41: string `username` has 12 occurrences, make it a constant (goconst)
		if setting.AuthProxyHeaderProperty == "username" {
		                                      ^

mjtrangoni committed:
pkg/api/*: Fix goconst issues with `GET` and `DELETE` strings

See,
$ golangci-lint run --disable-all --deadline 10m --enable=goconst
pkg/api/http_server.go:246:23: string `GET` has 3 occurrences, make it a constant (goconst)
	if ctx.Req.Method != "GET" || ctx.Req.URL.Path != "/metrics" {
	                     ^
pkg/api/playlist.go:36:48: string `DELETE` has 3 occurrences, make it a constant (goconst)
	if len(items) == 0 && c.Context.Req.Method != "DELETE" {
	                                              ^

mjtrangoni committed:
pkg/*: Fix all goconst issues

See,
$ golangci-lint run --disable-all --deadline 10m --enable=goconst
pkg/api/pluginproxy/ds_proxy_test.go:62:24: string `password` has 3 occurrences, make it a constant (goconst)
			setting.SecretKey = "password"
			                    ^
pkg/components/dynmap/dynmap_test.go:83:19: string `Street 42` has 4 occurrences, make it a constant (goconst)
	assert.True(s == "Street 42" && err == nil, "street should match")
	                 ^
pkg/components/dynmap/dynmap_test.go:71:19: string `anton` has 3 occurrences, make it a constant (goconst)
	assert.True(s == "anton" && err == nil, "name should be a string")
	                 ^
pkg/middleware/middleware_test.go:331:33: string `192.168.1.1, 2001::23` has 3 occurrences, make it a constant (goconst)
			setting.AuthProxyWhitelist = "192.168.1.1, 2001::23"
			                             ^
pkg/middleware/middleware_test.go:348:33: string `192.168.1.0/24, 2001::0/120` has 4 occurrences, make it a constant (goconst)
			setting.AuthProxyWhitelist = "192.168.1.0/24, 2001::0/120"
			                             ^
pkg/models/datasource_cache_test.go:45:23: string `password` has 3 occurrences, make it a constant (goconst)
		setting.SecretKey = "password"
		                    ^

mjtrangoni committed:
pkg/*: Fix trivial unconvert and ineffassign issues.

See,
$ golangci-lint run --disable-all --deadline 10m --enable=goconst --enable=deadcode --enable=gofmt --enable=ineffassign --enable=structcheck --enable=unconvert --enable=varcheck
pkg/services/quota/quota.go:51:13: unnecessary conversion (unconvert)
				if int64(usedSessions) > scope.DefaultLimit {
				        ^
pkg/api/login.go:166:44: unnecessary conversion (unconvert)
	decryptedError, err := util.Decrypt([]byte(decoded), setting.SecretKey)
	                                          ^
pkg/services/auth/auth_token.go:152:2: ineffectual assignment to `needsRotation` (ineffassign)
	needsRotation := false
	^

mjtrangoni committed:
Switch from gometalinter to golangci-lint

See https://github.com/golangci/golangci-lint,

 - GolangCI-Lint is a linters aggregator.
 - It's fast: on average 5 times faster than gometalinter.
 - It's easy to integrate and use, has nice output and has a
   minimum number of false positives.
 - It supports go modules.

bergquist commented on pkg/api/alerting_test.go:
```diff
@@ -66,7 +66,7 @@ func TestAlertingApiEndpoint(t *testing.T) {
 		})
 
-		loggedInUserScenarioWithRole("When calling GET on", "GET", "/api/alerts?dashboardId=1", "/api/alerts", m.ROLE_EDITOR, func(sc *scenarioContext) {
+		loggedInUserScenarioWithRole("When calling GET on", getMethod, "/api/alerts?dashboardId=1", "/api/alerts", m.ROLE_EDITOR, func(sc *scenarioContext) {
```
Please use `http.MethodGet` instead of a custom variable.

mjtrangoni replied:
Good catch! Thanks!

bergquist commented on .circleci/config.yml:
```diff
@@ -82,8 +82,8 @@ jobs:
       - run:
-          name: Gometalinter tests
-          command: './scripts/gometalinter.sh'
+          name: linters tests
```
`linters` might be a bit confusing since we have both backend and frontend linters.

bergquist replied:
How about backendlinters?

mjtrangoni replied:
@bergquist Good point. I renamed it to `linters-backend`, similar to `test-frontend` and `test-backend`.

torkelo commented:
Can you update this? looks close to mergable

mjtrangoni commented:
@torkelo sure, sorry for the delay.

mjtrangoni committed:
move linters to linters-backend

mjtrangoni committed:
pkg/api/*: Use net/http constants instead of own const.

See, https://golang.org/pkg/net/http/#pkg-constants

mjtrangoni committed:
scripts/linters-backend.sh: Remove unnecessary installations

golangci-lint says,
```
All linters are vendored in the /vendor folder: their version is fixed,
they are builtin and you don't need to install them separately.
```

mjtrangoni commented:
@torkelo done, and rebased against master branch. I see this `go vet` issue, which makes the CI to fail local.

```
$ go vet ./pkg/...
# github.com/grafana/grafana/pkg/models
pkg/models/dashboard_version.go:36:2: struct field CreatedBy repeats json tag "createdBy" also at dashboard_version.go:25
```

torkelo commented:
that looks like a bug, bad code to have duplicate property name like that.

mjtrangoni commented:
@torkelo This is working now, and with `go1.12` vet.

bergquist commented:
We dont use 1.12 yet since `goconvey` is broken in 1.12

mjtrangoni commented:
@bergquist Thanks! @torkelo This would be ready to merge then.

bergquist commented:
I dont like those `goconst` changes. 
replacing all strings like that I tests makes it harder to understand the code. Any chance you can revert those changes? except that I think we should merge this.

bergquist commented:
Big thank you for taking the time to contribute to Grafana and Im sorry my feedback in this case was annoying since I asked you to change back again. Ive partially replaced gometalinter with golangci-lint in this pr https://github.com/grafana/grafana/pull/16610 I created a new PR since this was getting way behind master. Im closing this PR since its not even further away from master.

Once again.... sorry for the bad feedback.

mjtrangoni commented:
Hi @bergquist,
No problem, I had less spare time last months. But I am always happy to contribute.
